terminus-ui/chip/src/chip-collection.component.spec.ts

```diff
@@ -132,44 +134,73 @@ describe(`TsChipCollection`, function() {
     expect(instance.value).toEqual(['banana']);
   });
 
+  describe(`empty`, () => {
+
+    test(`should check the number of chips if the collection exists`, () => {
+      fixture = createComponent(testComponents.NoChip);
+      chipCollectionInstance = getChipCollectionInstance(fixture);
+      fixture.detectChanges();
+
+      expect(chipCollectionInstance.empty).toEqual(true);
+    });
+
+  });
+
   test(`should be able to set isSelectable from chip collection and populate to chips`, fakeAsync(function() {
-    setupStandardCollection('RegularChip');
+    setupStandardCollection(testComponents.RegularChip);
 
     chipCollectionInstance.isSelectable = true;
     const chip = chips.toArray()[0];
     fixture.detectChanges();
+
     chipInstance.select();
     tick(1000);
     fixture.detectChanges();
 
     expect(chip.chipCollectionSelectable).toBeTruthy();
   }));
 
-  test(`should emit remove event when chip is removed`, fakeAsync(function() {
-    setupStandardCollection('RegularChip');
-    const chip = fixture.debugElement.query(By.css('.ts-chip')).nativeElement;
-    const chipRemovalButton = chip.querySelector('.ts-chip-remove');
-    chipCollectionInstance.remove = jest.fn();
-    chipCollectionInstance.removed.subscribe(event => {
-      expect(event).toEqual('apple');
-    });
+  test(`should emit 'collectionChange' and 'removed' events when a chip is removed`, function() {
+    setupStandardCollection(testComponents.Events);
+    const buttons = fixture.nativeElement.querySelectorAll('.c-chip__remove');
+    const chipRemovalButton = buttons[1];
+    fixture.componentInstance.change = jest.fn();
+    fixture.componentInstance.removed = jest.fn();
 
-    dispatchMouseEvent(chipRemovalButton, 'click');
-    tick(1000);
+    expect(chipCollectionInstance.chips.length).toEqual(2);
+
+    chipRemovalButton.click();
+    fixture.detectChanges();
+
+    expect(fixture.componentInstance.change.mock.calls[0][0].source.chips.length).toEqual(1);
+    expect(fixture.componentInstance.removed.mock.calls[0][0].chip.value).toEqual('banana');
+    expect.assertions(3);
+  });
+
+  test(`should emit the tabUpdateFocus event`, fakeAsync(() => {
+    setupStandardCollection(testComponents.Events);
+    fixture.componentInstance.tabbed = jest.fn();
+    const event = document.createEvent('KeyboardEvent');
+    event.initEvent('keydown', true, false);
+    Object.defineProperties(event, { code: { get: () => KEYS.TAB.code } });
+    Object.defineProperties(event, { keyCode: { get: () => KEYS.TAB.keyCode } });
```

Comment:
We've been using this `.createEvent`, `.initEvent` and `Object.defineProperties` quite a bit to define a keyboard event... maybe could extract that out to ngx-tools? just a thought...

Reply:
yeah, the issue is that we already have this in the tools library, but seem to hit odd issues where we need to create our own. I haven't pinpointed what happens so I'm not sure what we'd need.

I did notice we are setting the `code` to both the `key` and the `code` here https://github.com/GetTerminus/ngx-tools/blob/release/ngx-tools/testing/src/utilities/events/create-keyboard-event.ts#L30-L32

And I noticed that the CDK was listening for the actual number code under the `keyCode` property. So possibly just adding that to the helper function may work.